### PR TITLE
[clang-format] Fix requires misannotation with comma

### LINKED ISSUE #65904: [clang-format] incorrect formatting with `&` qualifier and requires clause with multi-arg template
The first one has no `&` qualifier and is formatted correct.

The second one adds `&` and clang-format gets confused, I don't think it's aware this is a function anymore.
```cpp
struct S {
    template <class... Args>
    constexpr void f()
        requires(true && C<true, Args...>)
    {
        return;
    }

    template <class... Args>
        constexpr void f() & requires(true && C<true, Args...>) { return; }
};
```

The `const` and `const&` qualifiers are parsed correctly. `&&` is also incorrect.

```cpp
struct S {
    template <class... Args>
    constexpr void f() const
        requires(true && C<true, Args...>)
    {
        return;
    }

    template <class... Args>
    constexpr void f() const&
        requires(true && C<true, Args...>)
    {
        return;
    }

    template <class... Args>
        constexpr void f() && requires(true && C<true, Args...>) { return; }
};
```

https://godbolt.org/z/qKTWec96e

### PR DESCRIPTION
**Pull request by rymiel**
clang-format uses a heuristic to determine if a requires() is either a requires clause or requires expression, based on what is in the parentheses. Part of this heuristic assumed that a requires clause can never contain a comma, however this is not the case if said comma is in the template argument of a type.

This patch allows commas to appear in a requires clause if an angle bracket `<` has been opened.

Fixes https://github.com/llvm/llvm-project/issues/65904

### CODE REVIEW

File: clang/lib/Format/UnwrappedLineParser.cpp [hunk 1]
```diff
@@ -3369,9 +3369,13 @@ bool clang::format::UnwrappedLineParser::parseRequires() {
...
-      FormatTok = Tokens->setPosition(StoredPosition);
-      parseRequiresExpression(RequiresToken);
-      return false;
+      if (OpenAngles == 0) {
+        FormatTok = Tokens->setPosition(StoredPosition);
+        parseRequiresExpression(RequiresToken);
+        return false;
+      } else {
+        break;
+      }
```
Comment (rymiel) [inline clang/lib/Format/UnwrappedLineParser.cpp:3378]: I realized after that i could do this. Would this be preferred?
```suggestion
      if (OpenAngles != 0)
        break;
      FormatTok = Tokens->setPosition(StoredPosition);
      parseRequiresExpression(RequiresToken);
      return false;
```

Comment (HazardyKnusperkeks) [reply clang/lib/Format/UnwrappedLineParser.cpp:3378]: I prefer the latter.

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 2]
```diff
@@ -1016,6 +1016,14 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresClausesAndConcepts) {
   ASSERT_EQ(Tokens.size(), 22u) << Tokens;
   EXPECT_TOKEN(Tokens[4], tok::kw_requires, TT_RequiresClause);
   EXPECT_TOKEN(Tokens[10], tok::ampamp, TT_PointerOrReference);
+
+  Tokens = annotate("void f() & requires(true) {}");
+  EXPECT_EQ(Tokens.size(), 12u) << Tokens;
+  EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
+
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:1023]: Check for the requires that it is a `TT_RequiresClause`.

File: clang/lib/Format/UnwrappedLineParser.cpp (hunk 1 above)
Comment (owenca) [reply clang/lib/Format/UnwrappedLineParser.cpp:3378]: The former is more consistent with the rest of the `switch` statement, except that there should be no `else` after the `return`. So I would prefer:
```
      if (OpenAngles == 0) {
        FormatTok = Tokens->setPosition(StoredPosition);
        parseRequiresExpression(RequiresToken);
        return false;
      }
      break;
```

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 3]
```diff
@@ -1016,6 +1016,14 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresClausesAndConcepts) {
...
+  Tokens = annotate("void f() & requires(C<true, true>) {}");
+  EXPECT_EQ(Tokens.size(), 17u) << Tokens;
+  EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
 }
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:1027]: Just to be sure, can you add a test with a requires expression like
``` c++
requires(C<true, true> c) {
{ c.foo(); }
}
```

**Review by owenca — APPROVED** [review]

**Review by HazardyKnusperkeks — CHANGES_REQUESTED** [review]

File: clang/lib/Format/UnwrappedLineParser.cpp (hunk 1 above)
Comment (HazardyKnusperkeks) [reply clang/lib/Format/UnwrappedLineParser.cpp:3378]: > The former is more consistent with the rest of the `switch` statement, except that there should be no `else` after the `return`. So I would prefer:
> 
> ```
>       if (OpenAngles == 0) {
>         FormatTok = Tokens->setPosition(StoredPosition);
>         parseRequiresExpression(RequiresToken);
>         return false;
>       }
>       break;
> ```

Also good, no `else` is my main point.

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 4]
```diff
@@ -1019,10 +1019,12 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresClausesAndConcepts) {
 
   Tokens = annotate("void f() & requires(true) {}");
   EXPECT_EQ(Tokens.size(), 12u) << Tokens;
+  EXPECT_TOKEN(Tokens[5], tok::kw_requires, TT_RequiresClause);
   EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
```
Comment (owenca) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:1023]:
```suggestion
  EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
  EXPECT_TOKEN(Tokens[5], tok::kw_requires, TT_RequiresClause);
```

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 5]
```diff
@@ -1019,10 +1019,12 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresClausesAndConcepts) {
...
 
   Tokens = annotate("void f() & requires(C<true, true>) {}");
   EXPECT_EQ(Tokens.size(), 17u) << Tokens;
+  EXPECT_TOKEN(Tokens[5], tok::kw_requires, TT_RequiresClause);
   EXPECT_TOKEN(Tokens[4], tok::amp, TT_PointerOrReference);
```
Comment (owenca) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:1028]: Ditto.

File: clang/unittests/Format/TokenAnnotatorTest.cpp [hunk 6]
```diff
@@ -1135,6 +1145,14 @@ TEST_F(TokenAnnotatorTest, UnderstandsRequiresExpressions) {
   EXPECT_TOKEN(Tokens[3], tok::kw_requires, TT_RequiresExpression);
   EXPECT_TOKEN(Tokens[4], tok::l_paren, TT_RequiresExpressionLParen);
   EXPECT_TOKEN(Tokens[14], tok::l_brace, TT_RequiresExpressionLBrace);
+
+  Tokens = annotate("bool foo = requires(C<true, true> c) {\n"
+                    "  { c.foo(); }\n"
+                    "}");
```
Comment (owenca) [inline clang/unittests/Format/TokenAnnotatorTest.cpp:1151]:
```suggestion
                    "};");
```

**Review by owenca — COMMENTED** [review]
Nits.

**Review by owenca — APPROVED** [review]

**Review by HazardyKnusperkeks — APPROVED** [review]

**Review by mydeveloperday — APPROVED** [review]

**Pull request merged**